### ISSUE TITLE
build: check-snapshot-anchors count_bare mis-scans macro text and // inside string literals

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#875]
## Summary

`count_bare()` in `check-snapshot-anchors.py` does not track string literals when scanning for `insta::assert_json_snapshot!(` calls, so the same `insta::assert_json_snapshot!(` text appearing inside a string body is mis-scanned — producing both a spurious bare count (false positive, blocks CI) and a missed real call (false negative, lets an unanchored snapshot through).

## Location

- `check-snapshot-anchors.py:185-208` (`count_bare`)
- specifically the line-comment guard at `:192-195` and the absence of any string-literal span check before counting a `MACRO_RE` match

## Evidence

`count_bare` filters matches two ways: it skips `/* … */` via `block_comment_spans`, and skips line comments via a naive substring test:

```python
line_start = source.rfind("\n", 0, macro_start) + 1
line_prefix = source[line_start:macro_start]
if "//" in line_prefix:
    continue
```

Neither path accounts for the match (or the `//`) being inside a Rust string literal. The module already has correct string-skipping scanners (`find_macro_call_end`, `block_comment_spans`), but `count_bare` does not reuse them to gate the match itself.

**False positive (spurious gate failure).** The literal macro text inside an inline-anchored snapshot body (or any fixture string containing Rust source) is counted as a *second*, bare call:

```rust
insta::assert_json_snapshot!(metric.x, @r###"
{ "note": "compare to insta::assert_json_snapshot!(other)" }
"###);
```

Driving this through the module's own logic yields `bare = 1` (expected `0`): the outer call is correctly anchored, but the occurrence inside the `@r###"…"###` body is counted separately.

**False negative (unanchored snapshot slips through).** A `//` inside a string literal earlier on the same line trips the `"//" in line_prefix` guard, so a genuinely bare snapshot is skipped entirely:

```rust
let url = "http://x"; insta::assert_json_snapshot!(metric.x);
```

Driving this through the module yields `bare = 0` (expected `1`).

Both reproducers were executed against the file's own functions (`MACRO_RE`, `block_comment_spans`, `find_macro_call_end`, `has_preceding_anchor`).

## Expected Behavior

A `MACRO_RE` match that falls inside a string literal must be ignored, and the line-comment guard must only treat a `//` as a comment when it is not inside a string. Anchored snapshots must not contribute spurious bare counts; real bare snapshots on a line containing a string with `//` must be counted.

## Actual Behavior

Macro text inside a string body inflates the bare count (spurious CI failure); a `//` inside a string before the macro suppresses a real bare count (gate bypass).

## Impact

This is a CI gate (`make snapshot-anchors`, the `lint` job, and the pre-commit hook). The false negative lets an unanchored snapshot regress the baseline undetected — the exact failure mode (#95, lesson 2) the gate exists to prevent. The false positive blocks CI on legitimate code; snapshot tests for a *code analyzer* routinely embed Rust source (containing `//` and macro-like text) as fixture strings, so both cases are realistic, not contrived. No current metric file triggers them, so this is latent.

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

`count_bare` scans raw source with `MACRO_RE.finditer` and gates each hit with (a) a `/* */` block-comment span check and (b) a naive `"//" in line_prefix` substring test. Neither accounts for the match — or the `//` — being inside a Rust string literal. The module already has a correct string-aware scanner inside `block_comment_spans` / `find_macro_call_end`; `count_bare` simply doesn't reuse it for the match itself.

### Implementation Steps

1. Factor the string/line-comment/block-comment skipping that `block_comment_spans` already performs into a single scanner that returns *all* spans to ignore: block-comment spans **and** string-literal spans (regular + raw). A small refactor of `block_comment_spans` can emit both, or add a sibling `string_literal_spans(source)`.
2. In `count_bare`, after finding a `MACRO_RE` match, skip it if `macro_start` falls inside any ignore span (block comment **or** string literal). This subsumes the false-positive case (macro text inside an inline body or fixture string).
3. Replace the `if "//" in line_prefix: continue` heuristic with a proper line-comment determination that ignores `//` occurring inside a string literal — e.g., compute line-comment spans with the same scanner and test membership, rather than a raw substring search. This fixes the false negative.

### Tests to Add/Update

- Add a unit test (pytest or a `__main__`-guarded self-check) feeding `count_bare`-equivalent input where the macro text appears inside an `@r###"…"###` body — assert count excludes it.
- Add a test with a bare snapshot on a line containing a string literal with `//` (e.g. `"http://x"`) — assert it is counted.
- Add a test with a real `//`-commented-out snapshot — assert it is NOT counted (preserve existing behavior).

### Verification

- [ ] `python3 check-snapshot-anchors.py --show` output unchanged for the current tree (no current file triggers either bug, so counts must stay identical).
- [ ] New unit tests pass.
- [ ] `make snapshot-anchors` still green.

**Comment by dekobon** [comment]
Fixed in commit `64173d25` (branch `worktree-agent-a1f97f28013f5d40d`, not yet pushed).

`block_comment_spans` was replaced by a single string-aware scanner `scan_ignore_spans`, which classifies line-comment, block-comment, and string-literal spans in one walk. `count_bare_in_source` now skips a `MACRO_RE` match that falls inside a string literal or block comment, and consults the precomputed line-comment spans instead of the naive `"//" in line_prefix` substring test.

A `--self-test` harness was added with both reproducers:
- macro text inside an `@r###"…"###` body → `bare=0` (was 1, false positive)
- `let url = "http://x"; insta::assert_json_snapshot!(metric.x);` → `bare=1` (was 0, false negative / gate bypass)

Verified via in-place perturbation that each case fails against the old logic and passes against the new. `--show` output is byte-identical on the current tree (no current metric file triggers either bug), and `make snapshot-anchors` stays green. Not closing per instruction.
